user_table not returning all results

**Describe the bug**

`select * from jira_user` doesn't display all records and returns an inconsistent number of records every time it is run (even without clearing the cache)

**Additional information**

- There is no error so you'd think everything worked properly
- `select count(*) from jira.jira_user` does display the correct number of users
- Adding `order by` causes a 429 error every time. Example:
`select * from jira.jira_user order by display_name`
`Error: table 'jira_user' column 'group_names' requires hydrate data from getUserGroups, which failed with error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429.`
- Reducing `MaxConcurrency` seems to fix the issue. I reduced it from 50 to 5 and consistently get all records.
- Running against a Jira cloud instance with a few hundred users

**Steampipe version (`steampipe -v`)**
Example: v0.16.4

**Plugin version (`steampipe plugin list`)**
Example: v0.6.1

Thanks for the assistance. On previous days I was able to replicate this on every query. This morning `select * from jira_user` worked several times in a row. However after a few clear/query cycles I was able to replicate the issue (results are incomplete but no error is returned). Here are the logs:

> 2022-10-31 13:39:42.466 UTC [ERROR] steampipe-plugin-jira.plugin: [ERROR] 1667223580111: jira_user.getUserGroups: api_error="429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429"
> 2022-10-31 13:39:42.466 UTC [ERROR] steampipe-plugin-jira.plugin: [ERROR] 1667223580111: jira_user.getUserGroups: api_error="429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429"
> 2022-10-31 13:39:42.466 UTC [ERROR] steampipe-plugin-jira.plugin: [ERROR] 1667223580111: callHydrate getUserGroups finished with error: 429
>  Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429
> 2022-10-31 13:39:42.466 UTC [ERROR] steampipe-plugin-jira.plugin: [ERROR] 1667223580111: callHydrate getUserGroups finished with error: 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429
> 2022-10-31 13:39:42.466 UTC [ERROR] steampipe-plugin-jira.plugin: [ERROR] 1667223580111: table 'jira_user' failed to get column data, connectionCallId jira-1667223580111: table 'jira_user' column 'group_names' requires hydrate data from getUserGroups, which failed with error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429.
> 2022-10-31 13:39:42.466 UTC [WARN]  steampipe-plugin-jira.plugin: [WARN]  1667223580111: QueryData StreamError table 'jira_user' column 'group_names' requires hydrate data from getUserGroups, which failed with error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429.
> 2022-10-31 13:39:42.466 UTC [ERROR] steampipe-plugin-jira.plugin: [ERROR] 1667223580111: jira_user.getUserGroups: api_error="429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429"
> 2022-10-31 13:39:42.466 UTC [ERROR] steampipe-plugin-jira.plugin: [ERROR] 1667223580111: jira_user.getUserGroups: api_error="429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429"
> 2022-10-31 13:39:42.466 UTC [WARN]  steampipe-plugin-jira.plugin: [WARN]  1667223580111: hydrate error chan select error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429
> 2022-10-31 13:39:42.466 UTC [WARN]  steampipe-plugin-jira.plugin: [WARN]  1667223580111: getRow failed with error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429
> 2022-10-31 13:39:42.466 UTC [WARN]  steampipe-plugin-jira.plugin: [WARN]  1667223580111: QueryData StreamError 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429 (jira-1667223580111)
> 2022-10-31 13:39:42.466 UTC [WARN]  steampipe-plugin-jira.plugin: [WARN]  1667223580111: hydrate error chan select error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429
> 2022-10-31 13:39:42.466 UTC [WARN]  steampipe-plugin-jira.plugin: [WARN]  1667223580111: getRow failed with error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429
> 2022-10-31 13:39:42.466 UTC [WARN]  steampipe-plugin-jira.plugin: [WARN]  1667223580111: QueryData StreamError 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429 (jira-1667223580111)
> 2022-10-31 13:39:42.466 UTC [ERROR] steampipe-plugin-jira.plugin: [ERROR] 1667223580111: callHydrate getUserGroups finished with error: 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429
> 2022-10-31 13:39:42.466 UTC [WARN]  steampipe-plugin-jira.plugin: [WARN]  1667223580111: streamRows for jira-1667223580111 - execution has failed (table 'jira_user' column 'group_names' requires hydrate data from getUserGroups, which failed with error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429.
> 2022-10-31 13:39:42.466 UTC [WARN]  steampipe-plugin-jira.plugin: [WARN]  1667223580111: executeForConnection jira returned error table 'jira_user' column 'group_names' requires hydrate data from getUserGroups, which failed with error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429.
> 2022-10-31 13:39:42.466 UTC [ERROR] steampipe-plugin-jira.plugin: [ERROR] 1667223580111: callHydrate getUserGroups finished with error: 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429
> 2022-10-31 13:39:42.466 UTC [ERROR] steampipe-plugin-jira.plugin: [ERROR] 1667223580111: table 'jira_user' failed to get column data, connectionCallId jira-1667223580111: table 'jira_user' column 'group_names' requires hydrate data from getUserGroups, which failed with error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429.
> 2022-10-31 13:39:42.466 UTC [WARN]  steampipe-plugin-jira.plugin: [WARN]  1667223580111: error channel received table 'jira_user' column 'group_names' requires hydrate data from getUserGroups, which failed with error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429.
> 2022-10-31 13:39:42.466 UTC [WARN]  steampipe-plugin-jira.plugin: [WARN]  1667223580111: QueryData StreamError table 'jira_user' column 'group_names' requires hydrate data from getUserGroups, which failed with error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429.
> 2022-10-31 13:39:42.466 UTC [WARN]  steampipe-plugin-jira.plugin: [WARN]  1667223580111: hydrate error chan select error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429
> 2022-10-31 13:39:42.466 UTC [WARN]  steampipe-plugin-jira.plugin: [WARN]  1667223580111: getRow failed with error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429
> 2022-10-31 13:39:42.466 UTC [WARN]  steampipe-plugin-jira.plugin: [WARN]  1667223580111: QueryData StreamError 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429 (jira-1667223580111)
> 2022-10-31 13:39:42.466 UTC [WARN]  steampipe-plugin-jira.plugin: [WARN]  1667223580111: hydrate error chan select error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429
> 2022-10-31 13:39:42.466 UTC [WARN]  steampipe-plugin-jira.plugin: [WARN]  1667223580111: getRow failed with error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429
> 2022-10-31 13:39:42.466 UTC [WARN]  steampipe-plugin-jira.plugin: [WARN]  1667223580111: QueryData StreamError 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429 (jira-1667223580111)
> 2022-10-31 13:39:42.466 UTC [ERROR] steampipe-plugin-jira.plugin: [ERROR] 1667223580111: table 'jira_user' failed to get column data, connectionCallId jira-1667223580111: table 'jira_user' column 'group_names' requires hydrate data from getUserGroups, which failed with error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429.
> 2022-10-31 13:39:42.466 UTC [WARN]  steampipe-plugin-jira.plugin: [WARN]  1667223580111: QueryData StreamError table 'jira_user' column 'group_names' requires hydrate data from getUserGroups, which failed with error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429.
> 2022-10-31 13:39:42.466 UTC [ERROR] steampipe-plugin-jira.plugin: [ERROR] 1667223580111: table 'jira_user' failed to get column data, connectionCallId jira-1667223580111: table 'jira_user' column 'group_names' requires hydrate data from getUserGroups, which failed with error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429.
> 2022-10-31 13:39:42.466 UTC [WARN]  steampipe-plugin-jira.plugin: [WARN]  1667223580111: QueryData StreamError table 'jira_user' column 'group_names' requires hydrate data from getUserGroups, which failed with error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429.

--- a/jira/errors.go
+++ b/jira/errors.go
@@ -2,7 +2,6 @@ package jira
 
 import (
 	"context"
-	"log"
 	"strings"
 
 	"github.com/turbot/steampipe-plugin-sdk/v4/plugin"
@@ -18,9 +17,9 @@ func isBadRequestError(err error) bool {
 
 func shouldRetryError(retryErrors []string) plugin.ErrorPredicateWithContext {
 	return func(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData, err error) bool {
-		
+
 		if strings.Contains(err.Error(), "429") {
-			log.Printf("[WARN] Received Rate Limit Error")
+			plugin.Logger(ctx).Warn("Received Rate Limit Error")
 			return true
 		}
 		return false

--- a/jira/errors.go
+++ b/jira/errors.go
@@ -19,7 +19,7 @@ func shouldRetryError(retryErrors []string) plugin.ErrorPredicateWithContext {
 	return func(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData, err error) bool {
 
 		if strings.Contains(err.Error(), "429") {
-			plugin.Logger(ctx).Warn("Received Rate Limit Error")
+			plugin.Logger(ctx).Debug("jira_errors.shouldRetryError", "rate_limit_error", err)
 			return true
 		}
 		return false

--- a/jira/errors.go
+++ b/jira/errors.go
@@ -1,6 +1,12 @@
 package jira
 
-import "strings"
+import (
+	"context"
+	"log"
+	"strings"
+
+	"github.com/turbot/steampipe-plugin-sdk/v4/plugin"
+)
 
 func isNotFoundError(err error) bool {
 	return strings.Contains(err.Error(), "404")
@@ -8,4 +14,15 @@ func isNotFoundError(err error) bool {
 
 func isBadRequestError(err error) bool {
 	return strings.Contains(err.Error(), "400")
+}
+
+func shouldRetryError(retryErrors []string) plugin.ErrorPredicateWithContext {
+	return func(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData, err error) bool {
+		
+		if strings.Contains(err.Error(), "429") {
+			log.Printf("[WARN] Received Rate Limit Error")
+			return true
+		}
+		return false
+	}
 }

--- a/jira/plugin.go
+++ b/jira/plugin.go
@@ -12,9 +12,9 @@ const pluginName = "steampipe-plugin-jira"
 // Plugin creates this (jira) plugin
 func Plugin(ctx context.Context) *plugin.Plugin {
 	p := &plugin.Plugin{
-		Name:               pluginName,
-		DefaultTransform:   transform.FromCamel(),
-		DefaultRetryConfig: &plugin.RetryConfig{ShouldRetryErrorFunc: shouldRetryError([]string{"429"})},
+		Name:               		pluginName,
+		DefaultTransform:   		transform.FromCamel(),
+		DefaultRetryConfig: 		&plugin.RetryConfig{ShouldRetryErrorFunc: shouldRetryError([]string{"429"})},
 		ConnectionConfigSchema: &plugin.ConnectionConfigSchema{
 			NewInstance: ConfigInstance,
 			Schema:      ConfigSchema,

--- a/jira/plugin.go
+++ b/jira/plugin.go
@@ -12,8 +12,9 @@ const pluginName = "steampipe-plugin-jira"
 // Plugin creates this (jira) plugin
 func Plugin(ctx context.Context) *plugin.Plugin {
 	p := &plugin.Plugin{
-		Name:             pluginName,
-		DefaultTransform: transform.FromCamel(),
+		Name:               pluginName,
+		DefaultTransform:   transform.FromCamel(),
+		DefaultRetryConfig: &plugin.RetryConfig{ShouldRetryErrorFunc: shouldRetryError([]string{"429"})},
 		ConnectionConfigSchema: &plugin.ConnectionConfigSchema{
 			NewInstance: ConfigInstance,
 			Schema:      ConfigSchema,

--- a/jira/table_jira_user.go
+++ b/jira/table_jira_user.go
@@ -26,6 +26,7 @@ func tableUser(_ context.Context) *plugin.Table {
 				// Limit concurrency to avoid a 429 too many requests error
 				Func:           getUserGroups,
 				MaxConcurrency: 50,
+			},
 		},
 		Columns: []*plugin.Column{
 			{

--- a/jira/table_jira_user.go
+++ b/jira/table_jira_user.go
@@ -26,8 +26,6 @@ func tableUser(_ context.Context) *plugin.Table {
 				// Limit concurrency to avoid a 429 too many requests error
 				Func:           getUserGroups,
 				MaxConcurrency: 50,
-				RetryConfig: 		&plugin.RetryConfig{ ShouldRetryErrorFunc: shouldRetryError([]string{"429"})},
-			},
 		},
 		Columns: []*plugin.Column{
 			{
@@ -162,7 +160,7 @@ func getUserGroups(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateDa
 		return nil, err
 	}
 
-	return  groups, nil
+	return groups, nil
 }
 
 //// TRANSFORM FUNCTION

--- a/jira/table_jira_user.go
+++ b/jira/table_jira_user.go
@@ -26,6 +26,7 @@ func tableUser(_ context.Context) *plugin.Table {
 				// Limit concurrency to avoid a 429 too many requests error
 				Func:           getUserGroups,
 				MaxConcurrency: 50,
+				RetryConfig: 		&plugin.RetryConfig{ ShouldRetryErrorFunc: shouldRetryError([]string{"429"})},
 			},
 		},
 		Columns: []*plugin.Column{
@@ -155,12 +156,13 @@ func getUserGroups(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateDa
 	}
 
 	groups, _, err := client.User.GetGroups(user.AccountID)
+
 	if err != nil {
 		plugin.Logger(ctx).Error("jira_user.getUserGroups", "api_error", err)
 		return nil, err
 	}
 
-	return groups, nil
+	return  groups, nil
 }
 
 //// TRANSFORM FUNCTION

